### DISCUSSION TITLE
Is there a way to override the histogram function?

### DISCUSSION THREAD

**Discussion by frosencrantz** [saulpw/visidata#1608] [Q&A]
How does one override the histogram function for the FreqSheet?

https://github.com/saulpw/visidata/blob/25a412a1b2fca7dd0f8e8ef06cc05aaf83b70e80/visidata/freqtbl.py#L55-L58

I was wondering how hard it would be to change the function to use the [Unicode block characterswe](https://en.wikipedia.org/wiki/Block_Elements) to get a smoother graph.

**Comment by andycraig** [comment]
I was able to override it by creating a plugin with the following code. When a histogram is created it adds the status message 'hello from histogram'. The code is a copy-pasted `visidata.freqtbl.FreqTableSheet.reload()`, although I did have to change the `super().initCols()` line to explicitly reference the parent class `PivotSheet`.

```python
from visidata import AttrColumn, asyncthread, Column, status, vlen, vd
from visidata.freqtbl import FreqTableSheet
from visidata.pivot import PivotSheet


@asyncthread
def reloadPatch(self):
    'Generate frequency table then reverse-sort by length.'
    PivotSheet.initCols(self) # changed from super().initCols()

    # add default bonus columns
    for c in [
        AttrColumn('count', 'sourcerows', type=vlen),
        Column('percent', type=float, getter=lambda col,row: len(row.sourcerows)*100/col.sheet.source.nRows),
    ]:
        self.addColumn(c)

    if self.options.disp_histolen and self.options.disp_histogram:
        def histogram(col, row):
            vd.status("hello from histogram")
            histogram = col.sheet.options.disp_histogram
            histolen = col.sheet.options.disp_histolen
            return histogram*(histolen*len(row.sourcerows)//col.sheet.largest)

        c = Column('histogram', type=str, getter=histogram, width=self.options.disp_histolen+2)
        self.addColumn(c)

    # two more threads
    vd.sync(self.addAggregateCols(),
            self.groupRows(self.updateLargest))

    if self.nCols > len(self.groupByCols)+3:  # hide percent/histogram if aggregations added
        self.column('percent').hide()
        self.column('histogram').hide()

    if not [c for c in self.groupByCols if vd.isNumeric(c)]:
        self.orderBy(self.column('count'), reverse=True)

FreqTableSheet.reload = reloadPatch
```

**Reply by frosencrantz** [reply]
I'll try this out thanks!

**Reply by frosencrantz** [reply]
@andycraig I hadn't read your post when I replied other than to see you had a solution and there was code.  When I looked at what you did I realized this is what I was trying to avoid by copying the rest of `visidata.freqtbl.FreqTableSheet.reload()`. Since Saul or Anja are likely to make changes that cause that code to be out-of-date.  

It doesn't look like `histogram` needs to be a sub-function in that method.   Is there a way to override that function without copying the rest of the surrounding method . .    If not I think perhaps VisiData could change so histogram is easier to override.

**Reply by andycraig** [reply]
@frosencrantz Yeah, it works but it's fragile.

It looks like it's possible although somewhat complicated to replace a nested function: https://stackoverflow.com/a/27550237/7785832 But I agree that it seems like `histogram` doesn't need to be a nested function so if it could be extracted outside of `reload` it feels like that would be more elegant.

**Reply by frosencrantz** [reply]
@andycraig would you be interested in fixing the code?  If you proposed a fix, I feel Saul and Anja would consider including it.

**Reply by andycraig** [reply]
I'd be happy to do that. I've created an issue to propose it: https://github.com/saulpw/visidata/issues/1621

**Reply by frosencrantz** [reply]
Thanks @andycraig.  I was thinking of a Pull Request, in case you want to directly contribute to VisiData.

Edit: Saul said that would be useful.  Thanks @andycraig!  Looking forward to your PR.
